### ISSUE TITLE
As a super admin, I want to email everyone in my organisation, so I can mobilise them.

### ISSUE THREAD

**Issue by camjackson** [rabblerouser/core#138]
**Background:**

Organisers want to get people to do stuff. Emailing them is a common way to communicate with them.

**User research:**

We know that people send emails to their members. Some people use e.g. mailchimp, some people do it through CRMs, some people have custom tools, some use a combination of the above.

**Visual design:**

**Todo:** Add something to the UI to show people how to do this. Suggested implementation right now is just a magical address that you can send emails to, and they get forwarded to all members. Therefore the UI for this might just be a `mailto:` link.

**Todo:** Organisers need to be able to know what they emailed and to whom. This could be e.g. "for a particular member, what emails did they get?", or "for a particular email blast, who did I send it to?".

**Technical notes:**
 - Magical email address which forwards emails to all members
 - Which service does this go in???
 - Authenticate the email by looking at the `from:` field. Only forward it if it comes from a super admin's email address
 - Use SES because its API is simple, but don't be too coupled to it
 - Send individual emails for each member, rather than a bulk bcc: email. This is because we know that soon we will send a slightly different email to every single member, where each email might have the member's name, and importantly unique tracking URLs.
 - Note the second point under 'visual design', where we need to know what emails were sent to whom. We need to be able to reconstruct that information from the event stream. Unsure of implementation, but this might mean e.g. putting events on the stream to record that email X was sent to member Y.

**Comment by camjackson** [comment]
***------------- NOTE: THIS DIAGRAM IS OUT OF DATE AND INCORRECT! LEAVING IT HERE FOR HISTORICAL PURPOSES ONLY-------------***

![uploads/f3c12a87-f1dd-41a6-8d29-5d2b3f25a7d1/Rabble Rouser group mailing.png](https://s3-us-west-2.amazonaws.com/prod.huboard.com/uploads%2Ff3c12a87-f1dd-41a6-8d29-5d2b3f25a7d1%2FRabble+Rouser+group+mailing.png)

***------------- NOTE: THIS DIAGRAM IS OUT OF DATE AND INCORRECT! LEAVING IT HERE FOR HISTORICAL PURPOSES ONLY-------------***

**Comment by camjackson** [comment]
Following on from that image, there are 3 sub-tasks for this that can probably be done in parallel, as long as people communicate :)

- #140: Receive the email and put the first event on
- #141: Receive the bulk request and build out the specific mailing list
- #142: Actually send the emails.

This isn't including anything to do with admins being able to see that the email was sent, which probably requires some more analysis/design first, which is where this card is on the board right now.

**Comment by camjackson** [comment]
Also a question - where should we authenticate the original email request? I.e. checking that the sender of the trigger email is an admin that we know about? I feel like it should be synchronous validation that happens *before* we hit the stream, rather than just blindly putting garbage onto the stream.

But that would mean that the email receiver thing would need to know the list of current admin email addresses, which probably means it needs to be a stateful service, rather than a lambda. Which is not really a problem, it's just a change from the diagram as we drew it this afternoon.

cc: @akjones @yearofthedan @alec-brooks @tomsulston because you were in the room when we discussed it :)

**Comment by camjackson** [comment]
Oh and another general technical point: **correlation IDs**. They are a thing that we should do.

**Comment by camjackson** [comment]
Revised design for the RR group mailer stack:

![rr mailing v2](https://cloud.githubusercontent.com/assets/1930451/25561480/fcb1d26c-2db0-11e7-85b0-9e73dada6d96.png)

Will need to re-think a couple of the issues we currently have open for this epic.

**Comment by camjackson** [comment]
Context for the previous picture:

@akjones and I were talking today about what's left for the mailer stack, and double checking if the previous design was right. In the previous design, there's a lambda publishing an event `"send-email-to-group"`, but that lambda has no way to validate the payload it receives (via SES) before doing so.

That would mean that any random person could spam emails to SES and DOS the event sausage. So we spoke about instead having the lambda function POST the email payload to the group mailer service directly. The group mailer being a proper stateful service listening to the stream, it can know about admins and groups and members, so it can synchronously validate the email request first, and avoid putting garbage onto the stream.

That's the only real change in the above diagram. The other addition is the post-email events that the mailer publishes (`"email-sent"`/`"email-failed"`), which has already been added to the mailer.

**Comment by camjackson** [comment]
Closing this in favour of #160, which is the new meta/epic-level issue for our first email milestone.
